Rework job food/alchohol benefits into livers

coiax opened the pull request:
:cl: coiax
refactor: Job specific benefits around food and drink have been moved
into the liver. For example, a security officer starts with a liver
that gains more of a benefit from donuts and Quad Sec. Steal that liver,
and digest donuts like a cop!
add: Medical Doctors, the CMO and the chaplain have a new skillchip, allowing them
to infer a person's career choice from examining their liver.
Observers also have access to this knowledge.
add: Livers from clowns, clownops and bananium golems now honk when squeezed.
fix: Fixed an oversight where clownops couldn't see each other as syndicate agents.
fix: Fixed the clown operative leader not having the same traits as other clownops.
/:cl:

A number of food and drinks react differently depending on your starting
job. However, that starting job is unchangable and also a borderline
"magic number".

By moving the "special behaviour" of something like security officers
eating donuts, or engineers losing radiation by drinking Screwdrivers,
into traits on the liver, this makes the "origin" of that behaviour more
clearly defined, rather than something that's attached to the mind of
the person. (For example, now if a wizard mindswaps into a Security
Officer, they too can now digest donuts good.)

Having this behaviour be partially visible to the more medically
inclined members of the station (like doctors, and the chaplain for
"entrails reading mystic" themes), means that a dismembered liver tells
a story to those who know how to read it.

Some jobs have more "benefits" than others, for example the only thing
that the liver of a Quartermaster gives them is a sense of inadequacy
when consuming royal carpet.

Clowns having livers that honk make them easier to identify, and plays
into the retconned "bike horns are clown livers lore"? Also, why not cut
out a clown's liver then honk them with it? You monster.

coiax committed:
Rework job food/alchohol benefits into livers

:cl: coiax
refactor: Job specific benefits around food and drink have been moved
into the liver. For example, a security officer starts with a liver
that gains more of a benefit from donuts and Quad Sec. Steal that liver,
and digest donuts like a cop!
add: Medical staff and the chaplain have a new skillchip, allowing them
to infer a person's career choice from examining their liver.
Observers also have access to this knowledge.
add: The liver from a clown now honks when squeezed.
/:cl:

A number of food and drinks react differently depending on your starting
job. However, that starting job is unchangable and also a borderline
"magic number".

By moving the "special behaviour" of something like security officers
eating donuts, or engineers losing radiation by drinking Screwdrivers,
into traits on the liver, this makes the "origin" of that behaviour more
clearly defined, rather than something that's attached to the mind of
the person. (For example, now if a wizard mindswaps into a Security
Officer, they too can now digest donuts good.)

Having this behaviour be partially visible to the more medically
inclined members of the station (like doctors, and the chaplain for
"entrails reading mystic" themes), means that a dismembered liver tells
a story to those who know how to read it.

Some jobs have more "benefits" than others, for example the only thing
that the liver of a Quartermaster gives them is a sense of inadequacy
when consuming royal carpet.

Clowns having livers that honk make them easier to identify, and plays
into the retconned "bike horns are clown livers lore"? Also, why not cut
out a clown's liver then honk them with it? You monster.

coiax committed:
Merge branch 'master' into the-liver-of-a-clown

coiax commented on code/modules/reagents/chemistry/reagents/other_reagents.dm:
```diff
@@ -2126,9 +2128,6 @@
 
-/datum/reagent/peaceborg
-	can_synth = FALSE
-
```
This change is not mentioned, but this type path isn't used anywhere, and all of its child reagents set `can_synth` to TRUE, so it can be safely removed?

Qustinnus commented on code/game/objects/items/food/pastries.dm:
```diff
@@ -56,7 +56,8 @@
 /obj/item/food/donut/proc/check_liked(fraction, mob/living/carbon/human/H)
-	if(HAS_TRAIT(H.mind, TRAIT_LAW_ENFORCEMENT_METABOLISM) && !HAS_TRAIT(H, TRAIT_AGEUSIA))
+	var/obj/item/organ/liver/liver = H.getorganslot(ORGAN_SLOT_LIVER)
+	if(!HAS_TRAIT(H, TRAIT_AGEUSIA) && liver && HAS_TRAIT(liver, TRAIT_LAW_ENFORCEMENT_METABOLISM))
```
its a bit weird that changing your liver affects your taste

jacklecroy replied:
So then this part could be put in the tongue or stomach?

coiax replied:
For another PR, could make a new "mind_trait" that just makes you like donuts, changing this. Or I could do this in the same one if that's not too complex.

Sparkezel commented:
How will that work if you get shot by a mutation toxin syringe or get turned into a fly by the teleporter?

ExcessiveUseOfCobblestone commented:
Remove from paramedic

coiax commented:
> 
> 
> Remove from paramedic

Remove the Entrails Reader skillchip? Because paramedics already had benefits from Doctor's Delight.

coiax committed:
Paramedics can no longer read entrails

coiax committed:
Merge branch 'master' into the-liver-of-a-clown

coiax committed:
Merge branch 'master' into the-liver-of-a-clown

coiax committed:
Remove entrail reader skillchip from virologist

Not a core medical worker, focused on viruses, etc.

coiax committed:
Tweak donut love to be in security mind

A love of donuts is not in your metabolism, but in your mind. Even if a
security officer's brain is put in another body, they will still love
donuts.

coiax committed:
Merge branch 'master' into the-liver-of-a-clown

coiax committed:
Removes merge markers and saves changes

Correctly made the changes to `drink_reagents.dm` but didn't save the
file before committing.

coiax committed:
Update clown ops and banananium golems liver

- Clown ops and bananium golems now have livers with "comedy
  metabolism"; thus removing the need for the "CLOWNLIKE" trait.
- Clown ops were not getting a HUD since the "apply_innate_effects"
  procs were not calling their parent.
- Moved the squeak component inside the liver, which is added when
  "COMEDY_METABOLISM" is added to the liver.

coiax committed:
Merge branch 'master' into the-liver-of-a-clown

Rohesie commented on code/modules/antagonists/nukeop/clownop.dm:
```diff
@@ -5,22 +5,28 @@
+	log_admin("[key_name(admin)] has clown op'ed [key_name(new_owner)].")
 
 /datum/antagonist/nukeop/clownop/apply_innate_effects(mob/living/mob_override)
+	..()
```
```suggestion
	. = ..()
```
Always preserve the return value (even if it's just `null`) unless you have a specific reason not to.

Rohesie commented on code/modules/antagonists/nukeop/clownop.dm:
```diff
@@ -5,22 +5,28 @@
+	ADD_TRAIT(L, TRAIT_NAIVE, CLOWNOP_TRAIT)
 
 /datum/antagonist/nukeop/clownop/remove_innate_effects(mob/living/mob_override)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/antagonists/nukeop/clownop.dm:
```diff
@@ -5,22 +5,28 @@
+
+/datum/antagonist/nukeop/clownop/equip_op()
+	. = ..()
+	var/mob/living/L = owner.current
+	var/obj/item/organ/liver/liver = L.getorganslot(ORGAN_SLOT_LIVER)
```
```suggestion
	var/mob/living/living_owner = owner.current
	var/obj/item/organ/liver/liver = living_owner.getorganslot(ORGAN_SLOT_LIVER)
```
Try to get used to informative variable names.

Rohesie commented on code/modules/antagonists/nukeop/clownop.dm:
```diff
@@ -29,8 +35,26 @@
+	challengeitem = /obj/item/nuclear_challenge/clownops
+
+/datum/antagonist/nukeop/leader/clownop/apply_innate_effects(mob/living/mob_override)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/antagonists/nukeop/clownop.dm:
```diff
@@ -29,8 +35,26 @@
+	ADD_TRAIT(L, TRAIT_NAIVE, CLOWNOP_TRAIT)
+
+/datum/antagonist/nukeop/leader/clownop/remove_innate_effects(mob/living/mob_override)
+	..()
```
```suggestion
	. = ..()
```

Rohesie commented on code/modules/jobs/job_types/_job.dm:
```diff
@@ -61,6 +61,10 @@
 
 	var/list/mind_traits // Traits added to the mind of the mob assigned this job
 
+	/// Traits added to the liver of the mob assigned this job
+	/// (used for the classic "cops heal from donuts" reaction, among others)
```
Check the multi-line comment format for autodocs.
Also don't fail to mention this is a lazy list.

Rohesie commented on code/modules/jobs/job_types/_job.dm:
```diff
@@ -87,6 +91,12 @@
 
+	var/obj/item/organ/liver/liver = H.getorganslot(ORGAN_SLOT_LIVER)
+
+	if(liver && liver_traits)
+		for(var/t in liver_traits)
```
```suggestion
	if(liver)
		for(var/t in liver_traits)
```
`for(var in null)` does what one would expect, no need to check if the list exists before

Rohesie commented on code/modules/mob/living/carbon/human/species_types/golems.dm:
```diff
@@ -532,7 +532,7 @@
 	fixed_mut_color = "ff0"
 	say_mod = "honks"
-	inherent_traits = list(TRAIT_RESISTHEAT,TRAIT_NOBREATH,TRAIT_RESISTCOLD,TRAIT_RESISTHIGHPRESSURE,TRAIT_RESISTLOWPRESSURE,TRAIT_NOFIRE,TRAIT_CHUNKYFINGERS, TRAIT_CLUMSY, TRAIT_CLOWNLIKE, TRAIT_RADIMMUNE,TRAIT_GENELESS,TRAIT_PIERCEIMMUNE,TRAIT_NODISMEMBER)
+	inherent_traits = list(TRAIT_RESISTHEAT, TRAIT_NOBREATH, TRAIT_RESISTCOLD, TRAIT_RESISTHIGHPRESSURE, TRAIT_RESISTLOWPRESSURE, TRAIT_NOFIRE, TRAIT_CHUNKYFINGERS, TRAIT_CLUMSY, TRAIT_RADIMMUNE, TRAIT_GENELESS, TRAIT_PIERCEIMMUNE, TRAIT_NODISMEMBER)
```
```suggestion
	inherent_traits = list(
		TRAIT_RESISTHEAT,
		TRAIT_NOBREATH,
		TRAIT_RESISTCOLD,
		TRAIT_RESISTHIGHPRESSURE,
		TRAIT_RESISTLOWPRESSURE,
		TRAIT_NOFIRE,
		TRAIT_CHUNKYFINGERS,
		TRAIT_CLUMSY,
		TRAIT_RADIMMUNE,
		TRAIT_GENELESS,
		TRAIT_PIERCEIMMUNE,
		TRAIT_NODISMEMBER,
		)
```
Feel free to disregard this suggestion if you don't feel it's more clear.

coiax replied:
I actually much prefer this style for uber long lists, but I've gotten in trouble before for breaking long lines into shorter ones.

Rohesie replied:
Who and where gave you trouble for something like this?

Rohesie commented on code/modules/surgery/organs/liver.dm:
```diff
@@ -21,6 +21,46 @@
+	// Would that make the clown more or less likely to honk it
+	AddComponent(/datum/component/squeak, list('sound/items/bikehorn.ogg'=1), 50, falloff_exponent = 20)
+
+/obj/item/organ/liver/proc/on_comedy_remove()
+	qdel(GetComponent(/datum/component/squeak))
```
Add an initialization parameter for the squeak component to register to signal removal of the comedy trait metabolism, and delete itself.
Avoid `GetComponent` for manipulating components.

coiax replied:
I'm not sure if I undestand what you want?

Add an initialization parameter to the squeak component, passing in a trait that it needs to watch the parent for, and then self delete when that trait is removed? That doesn't really seem like an improvement in code quality, which is why I feel like I've misunderstood it.

Rohesie replied:
The squeak component registers to signals based on the parent type.
You'll want to register to the comedy removal here, for liver types, and self-delete.
Components should handle their own deletion.

Rohesie commented on code/game/objects/items/clown_items.dm:
```diff
@@ -171,7 +171,7 @@
 /obj/item/bikehorn
 	name = "bike horn"
-	desc = "A horn off of a bicycle."
+	desc = "A horn off of a bicycle. Rumour has it that they're made from recycled clowns..."
```
```suggestion
	desc = "A horn off of a bicycle. Rumour has it that they're made from recycled clowns."
```

coiax committed:
Merge branch 'master' into the-liver-of-a-clown

coiax committed:
Clean variable names, comments, lists, descs

coiax committed:
Merge branch 'master' into the-liver-of-a-clown

coiax committed:
Move squeak component cleanup to component

coiax committed:
Update skillchip definition

Rohesie commented on code/modules/antagonists/nukeop/clownop.dm:
```diff
@@ -29,8 +35,26 @@
+	var/mob/living/L = owner.current || mob_override
+	ADD_TRAIT(L, TRAIT_NAIVE, CLOWNOP_TRAIT)
+
+/datum/antagonist/nukeop/leader/clownop/remove_innate_effects(mob/living/mob_override)
+	. = ..()
+	var/mob/living/L = owner.current || mob_override
+	REMOVE_TRAIT(L, TRAIT_NAIVE, CLOWNOP_TRAIT)
+
```
```suggestion
/datum/antagonist/nukeop/leader/clownop/remove_innate_effects(mob/living/mob_override)
	var/mob/living/L = owner.current || mob_override
	REMOVE_TRAIT(L, TRAIT_NAIVE, CLOWNOP_TRAIT)
	return ..()

```
Wrapping-up procs should in general act before calling parent.

Rohesie commented on code/datums/components/squeak.dm:
```diff
@@ -39,6 +39,10 @@
 
+	if(istype(parent, /obj/item/organ/liver))
+		// Liver squeaking is depending on them functioning like a clown's liver
+		RegisterSignal(parent, SIGNAL_REMOVETRAIT(TRAIT_COMEDY_METABOLISM),		.proc/cleanup)
```
```suggestion
		RegisterSignal(parent, SIGNAL_REMOVETRAIT(TRAIT_COMEDY_METABOLISM), .proc/cleanup)
```

Rohesie commented on code/datums/components/squeak.dm:
```diff
@@ -116,3 +120,8 @@
 	//If the dir changes it means we're going through a bend in the pipes, let's pretend we bumped the wall
 	if(old_dir != new_dir)
 		play_squeak()
+
+/datum/component/squeak/proc/cleanup(datum/source, trait)
```
```suggestion

///Called when the comedy liver trait is removed, to stop it from making honking sounds.
/datum/component/squeak/proc/on_comedy_trait_removal(datum/source, trait)
```

Rohesie commented on code/modules/surgery/organs/liver.dm:
```diff
@@ -21,6 +21,44 @@
+	. = ..()
+	// If the liver handles foods like a clown, it honks like a bike horn
+	// Don't think about it too much.
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_COMEDY_METABOLISM), .proc/on_comedy)
+
+/obj/item/organ/liver/proc/on_comedy()
+	// Are clown "bike" horns made from the livers of ex-clowns?
```
Could you document this proc, maybe give it a more informative name?
Add the signal handler macro as well.

coiax commented:
@Rohesie It feels REALLY weird adding the component in the liver by listening to a trait addition, but having the component handle its own removal, by listening to the trait removal in the same liver. Are you sure this is the best way?

Qustinnus commented:
that is the right way

coiax committed:
Merge branch 'master' into the-liver-of-a-clown

coiax committed:
Change function names, add documentation

coiax committed:
Add SIGNAL_HANDLER as well

Rohesie commented on code/modules/antagonists/nukeop/clownop.dm:
```diff
@@ -29,8 +35,26 @@
+/datum/antagonist/nukeop/leader/clownop/remove_innate_effects(mob/living/mob_override)
+	var/mob/living/L = owner.current || mob_override
+	REMOVE_TRAIT(L, TRAIT_NAIVE, CLOWNOP_TRAIT)
+	. = ..()
```
```suggestion
	return ..()
```

Rohesie commented on code/modules/antagonists/nukeop/clownop.dm:
```diff
@@ -5,22 +5,28 @@
-	REMOVE_TRAIT(L, TRAIT_NAIVE, "clownop")
-	REMOVE_TRAIT(L, TRAIT_CLOWNLIKE, "clownop")
+	REMOVE_TRAIT(L, TRAIT_NAIVE, CLOWNOP_TRAIT)
+	. = ..()
```
```suggestion
	return ..()
```

Commit:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>
